pkg/provisioning/provision.go

```diff
@@ -0,0 +1,353 @@
+		if err != nil {
+			return nil, nil, err
+		}
+		destMap[k] = destState
+		srcMap[k] = srcState
+	}
+
+	err := s.deletePipeline(ctx, pl, config)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return destMap, srcMap, nil
+}
+
+func (s *Service) deletePipeline(ctx context.Context, pl *pipeline.Instance, config PipelineConfig) error {
+	var err error
+	// remove pipeline processors
+	for k := range config.Processors {
+		_, err = s.pipelineService.RemoveProcessor(ctx, pl, k)
+		if err != nil {
+			return err
+		}
+		err = s.processorService.Delete(ctx, k)
+		if err != nil {
+			return err
+		}
+	}
+	// remove connector processors
+	for k, cfg := range config.Connectors {
+		for k1 := range cfg.Processors {
+			_, err = s.connectorService.RemoveProcessor(ctx, k, k1)
+			if err != nil {
+				return err
+			}
+			err = s.processorService.Delete(ctx, k1)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	// remove connectors
+	for k := range config.Connectors {
+		_, err = s.pipelineService.RemoveConnector(ctx, pl, k)
+		if err != nil {
+			return err
+		}
+		err = s.connectorService.Delete(ctx, k)
+		if err != nil {
+			return err
+		}
+	}
+	err = s.pipelineService.Delete(ctx, pl)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *Service) createPipeline(ctx context.Context, id string, config PipelineConfig) (*pipeline.Instance, error) {
+	cfg := pipeline.Config{
+		Name:        config.Name,
+		Description: config.Description,
+	}
+
+	pl, err := s.pipelineService.Create(ctx, id, cfg, pipeline.ProvisionTypeConfig)
+	if err != nil {
+		return nil, err
+	}
+	return pl, nil
+}
+
+func (s *Service) createConnector(ctx context.Context, pipelineID string, id string, config ConnectorConfig, destMap map[string]connector.DestinationState, srcMap map[string]connector.SourceState) error {
+	cfg := connector.Config{
+		Name:       config.Name,
+		Plugin:     config.Plugin,
+		PipelineID: pipelineID,
+		Settings:   config.Settings,
+	}
+
+	connType := connector.TypeSource
+	if config.Type == TypeDestination {
+		connType = connector.TypeDestination
+	}
+
+	conn, err := s.connectorService.Create(ctx, id, connType, cfg, connector.ProvisionTypeConfig)
+	if err != nil {
+		return cerrors.Errorf("could not create connector %q on pipeline %q: %w", id, pipelineID, err)
+	}
+
+	// check if states should be updated
+	if len(srcMap) != 0 || len(destMap) != 0 {
+		switch v := conn.(type) {
+		case connector.Destination:
+			v.SetState(destMap[id])
+		case connector.Source:
+			v.SetState(srcMap[id])
+		}
+	}
+
+	return nil
+}
+
+func (s *Service) getConnectorState(ctx context.Context, id string, plugin string) (connector.DestinationState, connector.SourceState, error) {
+	var destState connector.DestinationState
+	var srcState connector.SourceState
+	conn, err := s.connectorService.Get(ctx, id)
+	if err != nil {
+		return destState, srcState, cerrors.Errorf("could not get connector %q: connector id cannot be changed", id)
```

Comment:
This doesn't seem right - if the connector doesn't exist the ID was probably changed, but that's a valid thing to do. In that case the user essentially added a new connector to the config and deleted the old one.